components/num-util/src/fixed_decimal.rs

```diff
@@ -699,32 +759,77 @@ fn test_lower_magnitude_bounds() {
     let dec_backup = dec.clone();
     assert_eq!(Error::Limit, dec.multiply_pow10(-1).unwrap_err());
     assert_eq!(dec, dec_backup, "Value should be unchanged on failure");
-}
 
-#[test]
-fn test_from_str_isize_limits() {
-    let dec_max: FixedDecimal = std::isize::MAX.into();
-    let input = dec_max.to_string();
-    let dec: FixedDecimal = FixedDecimal::from_str(&input).unwrap();
-    assert_eq!(dec.to_string(), input);
-    assert_eq!(dec.write_len(), input.len());
-    let dec_min: FixedDecimal = std::isize::MIN.into();
-    let input = dec_min.to_string();
+    // Checking from_str for dec (which is valid)
+    let input = dec.to_string();
     let dec: FixedDecimal = FixedDecimal::from_str(&input).unwrap();
     assert_eq!(dec.to_string(), input);
     assert_eq!(dec.write_len(), input.len());
-}
 
-#[test]
-fn test_from_str_u128_limits() {
-    let dec_max: FixedDecimal = std::u128::MAX.into();
-    let input = dec_max.to_string();
-    let dec: FixedDecimal = FixedDecimal::from_str(&input).unwrap();
-    assert_eq!(dec.to_string(), input);
-    assert_eq!(dec.write_len(), input.len());
-    let dec_min: FixedDecimal = std::u128::MIN.into();
-    let input = dec_min.to_string();
-    let dec: FixedDecimal = FixedDecimal::from_str(&input).unwrap();
-    assert_eq!(dec.to_string(), input);
-    assert_eq!(dec.write_len(), input.len());
+    // The following tests are for testing strings that have sligthly less, sligthtly more, or exactly the extreme
+    // possible length (i16::MAX).
+    // Exactly i16::MAX zeros (no dot)
+    let alignment = 32768;
+    let input_str = format!("{:0fill$}", 0, fill = alignment);
+    let dec: FixedDecimal = FixedDecimal::from_str(&input_str).unwrap();
+    assert_eq!(dec.to_string(), input_str);
+    assert_eq!(dec.write_len(), input_str.len());
+
+    // Slightly less than i16::MAX zeros (no dot)
```

Comment:
There is code duplication in this function.  This is both good and bad.

The bad thing is that we generally keep code DRY ("don't repeat yourself"), so it seems like you would put the related logic into a helper function or a loop.

The good thing is that it helps pinpoint the exact assert that failed based on the line number.  You lose that in a for loop.  However, you can recover context information by passing an additional format string to the end of `assert_eq!`.

So, I would probably use a for loop.  Consider something like this (untested):

```rust
    #[derive(Debug)]
    struct TestCase {
        pub zeros_before_dot: usize,
        pub zeros_after_dot: usize,
        pub expected_err: Option<Error>,
    };
    let cases = [
        TestCase {
            zeros_before_dot: 32768,
            zeros_after_dot: 0,
            None,
        },
        TestCase {
            zeros_before_dot: 32769,
            zeros_after_dot: 0,
            Some(Error::Limit),
        },
        // ...
    };
    for cas in &cases {
        let mut input_str = format!("{:0fill$}", 0, fill = cas.zeros_before_dot);
        if cas.zeros_after_dot > 0 {
            input_str.push_str(".");
            input_str.push_str(&format!("{:0fill$}", 0, fill = cas.zeros_after_dot));
        }
        match input_str.parse() {
            case Ok(dec) => {
                assert_eq!(cas.expected_err, None, "{:?}", cas);
                assert_eq!(input_str, dec.to_string(), "{:?}", cas);
            },
            case Err(err) => {
                assert_eq!(cas.expected_err, Some(err), "{:?}", cas);
            },
        }
    }
```